universe/ValueRef.h

```diff
@@ -86,4 +103,91 @@ enum StatisticType : int {
 
 }
 
+//! Holds all FreeOrion named ValueRef%s.  ValueRef%s may be looked up by name.
+class NamedValueRefManager {
+public:
+    //using container_type = std::map<const std::string, const std::unique_ptr<ValueRef::AnyValueRef>>;
+    using key_type = std::string;
+    using value_type = std::unique_ptr<ValueRef::AnyValueRef>;
+    using container_type = std::map<key_type, value_type>;
+
+    using iterator = container_type::const_iterator;
+
+    //! Returns the ValueRef with the name @p name or nullptr if there is nov ValueRef with such a name or of the wrong type
+    //! use the free function GetValueRef(...) instead, mainly to save some typing.
+    template <typename T>
+    auto GetValueRef(const std::string& name) -> ValueRef::ValueRef<T>*;
+
+    //! Returns the ValueRef with the name @p name; you should use the
+    //! free function GetValueRef(...) instead, mainly to save some typing.
+    auto GetAnyValueRef(const std::string& name) const -> ValueRef::AnyValueRef*;
+
+    auto NumNamedValueRefs() const -> std::size_t { return m_value_refs.size(); }
+
+    //! iterator to the first value ref
+    FO_COMMON_API auto begin() const -> iterator;
+
+    //! iterator to the last + 1th value ref
+    FO_COMMON_API auto end() const -> iterator;
+
+    // Singleton
+    NamedValueRefManager&  operator=(NamedValueRefManager const&) =delete; // no copy via assignment
+
+    NamedValueRefManager(NamedValueRefManager const&) =delete;            // no copies via construction
+
+    ~NamedValueRefManager();
+
+    /** Sets named value refs to the value of \p future. */
+    FO_COMMON_API void SetNamedValueRefs(Pending::Pending<container_type>&& future);
+
+    //! Returns the instance of this singleton class; you should use the free
+    //! function GetNamedValueRefManager() instead
+    static NamedValueRefManager& GetNamedValueRefManager();
+
+    //! Returns a number, calculated from the contained data, which should be
+    //! different for different contained data, and must be the same for
+    //! the same contained data, and must be the same on different platforms
+    //! and executions of the program and the function. Useful to verify that
+    //! the parsed content is consistent without sending it all between
+    //! clients and server.
+    auto GetCheckSum() const -> unsigned int;
+
+    //! Register the @p value_ref under the evaluated @p name.
+    template <typename T>
+    std::string RegisterValueRef(std::string name, std::unique_ptr<T> vref);
+
+private:
+    NamedValueRefManager();
+
+    /** Assigns any m_pending_refs to m_value_refs. */
+    void CheckPendingRefs() const;
+
+    /** Future named value refs being parsed by parser.  mutable so that it can
+        be assigned to m_species_types when completed.*/
+    mutable boost::optional<Pending::Pending<container_type>> m_pending_refs = boost::none;
+
+    //! Map of ValueRef%s identified by a name
+    container_type m_value_refs;
```

Comment:
could also have separate containers for each type of valueref that can be stored and returned, to avoid needing to cast and to limit the search pool when the type is known.